include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,275 @@
+/// by Triton.
+/// \param minor Returns the TRITONCACHE API minor version supported
+/// by Triton.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_ApiVersion(
+    uint32_t* major, uint32_t* minor);
+
+///
+/// CacheEntry Lifetime Management
+///
+
+/// Create a new cache entry object. The caller takes ownership of the
+/// TRITONCACHE_CacheEntry object and must call
+/// TRITONCACHE_CacheEntryDelete to release the object.
+///
+/// \param entry Returns the new cache entry object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryNew(
+    TRITONCACHE_CacheEntry** entry);
+
+/// Delete a cache entry object.
+///
+/// \param entry The cache entry object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryDelete(
+    TRITONCACHE_CacheEntry* entry);
+
+///
+/// CacheEntry Field Management
+///
+
+// TODO: Proper API descriptions, lifetimes, copies, etc.
+
+// Get number of items in entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemCount(
+    TRITONCACHE_CacheEntry* entry, size_t* count);
+
+// Adds item to entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryAddItem(
+    TRITONCACHE_CacheEntry* entry, TRITONCACHE_CacheEntryItem* item);
+
+// Gets item at index from entry where 0 <= index < count and
+// 'count' is the value returned by TRITONCACHE_CacheEntryItemCount
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryGetItem(
+    TRITONCACHE_CacheEntry* entry, size_t index,
+    TRITONCACHE_CacheEntryItem** item);
+
+
+///
+/// CacheEntryItem Lifetime Management
+///
+
+/// Create a new cache entry item object. The caller takes ownership of the
+/// TRITONCACHE_CacheEntryItem object and must call
+/// TRITONCACHE_CacheEntryItemDelete to release the object.
+///
+/// \param item Returns the new cache entry item object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemNew(
+    TRITONCACHE_CacheEntryItem** item);
+
+/// Delete a cache entry item object.
+///
+/// \param item The cache entry item object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemDelete(
+    TRITONCACHE_CacheEntryItem* item);
+
+///
+/// CacheEntryItem Field Management
+///
+
+// TODO: Proper API descriptions, lifetimes, copies, etc.
+
+// Returns number of buffers held by item
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemBufferCount(
+    TRITONCACHE_CacheEntryItem* item, size_t* count);
+
+// Adds buffer to item
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemAddBuffer(
+    TRITONCACHE_CacheEntryItem* item, const void* base, size_t byte_size);
```

Comment:
Does the buffer API need to be memory type aware? For example, in the future we may decide to add support for GPU buffers. It seems like there is no way right now to specify what memory that the `base` pointer is in.

Reply:
This is not currently memory type aware. Currently everything would have to go through CPU memory. I agree it would be better if we included the fields to support other memory types for the API, even if they weren't completely supported yet.

To avoid too many single large PRs, I want to gather the changes in a side `rmccormick-cache-main` branch, When we sort out the final desired APIs in terms of entries/items/buffers, I think I can make this a follow-up PR so there's less to review.

Reply:
Thanks. Resolving.